File: apps/desktop/src-tauri/src/commands/licensing.rs
//! Tauri commands for licensing.

use crate::licensing;

/// Get the current app status (personal, commercial, or expired).
#[tauri::command]
pub fn get_license_status(app: tauri::AppHandle) -> licensing::AppStatus {
    licensing::get_app_status(&app)
}

/// Get the window title based on current license status.
#[tauri::command]
pub fn get_window_title(app: tauri::AppHandle) -> String {
    let status = licensing::get_app_status(&app);
    licensing::get_window_title(&status)
}

/// Activate a license key or short code (verify + commit in one call).
/// If the input is a short code (CMDR-XXXX-XXXX-XXXX), it first exchanges it for the full key.
/// Kept for backward compatibility — new code should use verify_license + commit_license.
#[tauri::command]
pub async fn activate_license(
    app: tauri::AppHandle,
    license_key: String,
) -> Result<licensing::LicenseInfo, licensing::LicenseActivationError> {
    licensing::activate_license_async(&app, &license_key).await
}

/// Verify a license key or short code without writing anything to disk.
/// Returns the verify result (LicenseInfo + full key) for the frontend to inspect
/// before deciding whether to commit.
#[tauri::command]
pub async fn verify_license(license_key: String) -> Result<licensing::VerifyResult, licensing::LicenseActivationError> {
    licensing::verify_license_async(&license_key).await
}

/// Persist a verified license key to disk and update caches.
/// Only call after verification confirms the key is valid.
#[tauri::command]
pub fn commit_license(
    app: tauri::AppHandle,
    license_key: String,
    short_code: Option<String>,
) -> Result<licensing::LicenseInfo, licensing::LicenseActivationError> {
    licensing::commit_license(&app, &license_key, short_code.as_deref())
}

/// Get information about the current license (if any).
#[tauri::command]
pub fn get_license_info(app: tauri::AppHandle) -> Option<licensing::LicenseInfo> {
    licensing::get_license_info(&app)
}

/// Mark the expiration modal as shown (so it won't show again).
#[tauri::command]
pub fn mark_expiration_modal_shown(app: tauri::AppHandle) {
    licensing::mark_expiration_modal_shown(&app);
}

/// Mark the commercial reminder as dismissed (resets the 30-day timer).
#[tauri::command]
pub fn mark_commercial_reminder_dismissed(app: tauri::AppHandle) {
    licensing::mark_commercial_reminder_dismissed(&app);
}

/// Reset license data (debug builds only).
#[tauri::command]
pub fn reset_license(app: tauri::AppHandle) {
    licensing::reset_license(&app);
}

/// Check if license needs re-validation with the server.
#[tauri::command]
pub fn needs_license_validation(app: tauri::AppHandle) -> bool {
    licensing::needs_validation(&app)
}

/// Check if a server validation has ever completed for the current license.
#[tauri::command]
pub fn has_license_been_validated(app: tauri::AppHandle) -> bool {
    licensing::has_been_validated(&app)
}

/// Validate license with server (async - call when needs_license_validation returns true).
/// If `transaction_id` is provided, uses it directly (for pre-commit validation).
/// If `None`, reads from the stored license (for periodic re-validation).
/// Returns Err on network/upstream errors so the frontend can distinguish from server rejection.
#[tauri::command]
pub async fn validate_license_with_server(
    app: tauri::AppHandle,
    transaction_id: Option<String>,
) -> Result<licensing::AppStatus, String> {
    licensing::validate_license_async(&app, transaction_id.as_deref()).await
}

File: apps/desktop/src/lib/tauri-commands/licensing.ts
// License commands

import { commands } from '$lib/ipc/bindings'
import { throwIpcError } from './ipc-types'

/** License types */
export type LicenseType = 'commercial_subscription' | 'commercial_perpetual'

/** Application license status */
export type LicenseStatus =
  | { type: 'personal'; showCommercialReminder: boolean }
  | { type: 'commercial'; licenseType: LicenseType; organizationName: string | null; expiresAt: string | null }
  | { type: 'expired'; organizationName: string | null; expiredAt: string; showModal: boolean }

/** License information from activation */
export interface LicenseInfo {
  email: string
  transactionId: string
  issuedAt: string
  organizationName: string | null
  licenseType: string | null
  shortCode: string | null
}

/** Result of verifying a license key without persisting it. */
export interface VerifyResult {
  info: LicenseInfo
  fullKey: string
  shortCode: string | null
}

/** Error codes from the license activation flow (matches Rust `LicenseActivationError` enum). */
export type LicenseActivationErrorCode =
  | 'invalidFormat'
  | 'badEncoding'
  | 'badSignature'
  | 'badPayload'
  | 'shortCodeNotFound'
  | 'networkError'
  | 'serverError'

/** Typed activation error returned by the backend. */
export interface LicenseActivationError {
  code: LicenseActivationErrorCode
  detail?: string
}

const validActivationErrorCodes = new Set<string>([
  'invalidFormat',
  'badEncoding',
  'badSignature',
  'badPayload',
  'shortCodeNotFound',
  'networkError',
  'serverError',
])

/**
 * Parses a Tauri invoke error into a typed activation error.
 *
 * Tauri 2 serializes `Serialize`-implementing error types directly as the rejection value.
 * The value may arrive as a parsed object or as a JSON string (depending on Tauri version/config).
 */
export function parseActivationError(e: unknown): LicenseActivationError | null {
  // Case 1: already a parsed object with a `code` field
  if (typeof e === 'object' && e !== null && 'code' in e) {
    const obj = e
    if (typeof obj.code === 'string' && validActivationErrorCodes.has(obj.code)) {
      return e as LicenseActivationError
    }
  }

  // Case 2: JSON string (Tauri may wrap the serialized error in a string)
  if (typeof e === 'string') {
    try {
      const parsed: unknown = JSON.parse(e)
      if (typeof parsed === 'object' && parsed !== null && 'code' in parsed) {
        const obj = parsed
        if (typeof obj.code === 'string' && validActivationErrorCodes.has(obj.code)) {
          return parsed as LicenseActivationError
        }
      }
    } catch {
      // Not JSON, fall through
    }
  }

  return null
}

/**
 * Gets the current application license status.
 * @returns Current license status (personal, commercial, or expired)
 */
export async function getLicenseStatus(): Promise<LicenseStatus> {
  return commands.getLicenseStatus()
}

/**
 * Gets the window title based on current license status.
 * @returns Window title string (like "Cmdr – Personal use only")
 */
export async function getWindowTitle(): Promise<string> {
  return commands.getWindowTitle()
}

/**
 * Activates a license key (verify + commit in one call).
 * Kept for backward compatibility. New code should use verifyLicense + commitLicense.
 */
export async function activateLicense(licenseKey: string): Promise<LicenseInfo> {
  const res = await commands.activateLicense(licenseKey)
  if (res.status === 'error') throwIpcError(res.error)
  return res.data
}

/** Verifies a license key offline without writing anything to disk. */
export async function verifyLicense(licenseKey: string): Promise<VerifyResult> {
  const res = await commands.verifyLicense(licenseKey)
  if (res.status === 'error') throwIpcError(res.error)
  return res.data
}

/** Persists a verified license key to disk and updates caches. */
export async function commitLicense(licenseKey: string, shortCode: string | null): Promise<LicenseInfo> {
  const res = await commands.commitLicense(licenseKey, shortCode)
  if (res.status === 'error') throwIpcError(res.error)
  return res.data
}

/**
 * Gets information about the current stored license.
 * @returns License info if a valid license is stored, null otherwise
 */
export async function getLicenseInfo(): Promise<LicenseInfo | null> {
  return commands.getLicenseInfo()
}

/**
 * Marks the expiration modal as shown to prevent showing it again.
 */
export async function markExpirationModalShown(): Promise<void> {
  await commands.markExpirationModalShown()
}

/**
 * Marks the commercial reminder as dismissed (resets the 30-day timer).
 */
export async function markCommercialReminderDismissed(): Promise<void> {
  await commands.markCommercialReminderDismissed()
}

/**
 * Resets all license data (debug builds only).
 */
export async function resetLicense(): Promise<void> {
  await commands.resetLicense()
}

/**
 * Checks if the license needs re-validation with the server.
 * Should be called on app startup to determine if validateLicenseWithServer should be invoked.
 * @returns True if validation is needed (7+ days since last validation)
 */
export async function needsLicenseValidation(): Promise<boolean> {
  return commands.needsLicenseValidation()
}

/**
 * Checks if a server validation has ever completed for the current license.
 * Returns false if the license was committed locally but never verified with the server.
 */
export async function hasLicenseBeenValidated(): Promise<boolean> {
  return commands.hasLicenseBeenValidated()
}

/**
 * Validates the license with the license server.
 * If transactionId is provided, uses it directly (for pre-commit validation).
 * If omitted, reads from the stored license (for periodic re-validation).
 */
export async function validateLicenseWithServer(transactionId?: string): Promise<LicenseStatus> {
  const res = await commands.validateLicenseWithServer(transactionId ?? null)
  if (res.status === 'error') throwIpcError(res.error)
  return res.data
}
